static/src/javascripts/projects/commercial/modules/dfp/on-slot-viewable.ts

```diff
@@ -1,38 +1,103 @@
+import type { AdSizeString } from '@guardian/commercial-core';
+import { adSizes } from '@guardian/commercial-core';
+import config from 'lib/config';
 import { getUrlVars } from '../../../../lib/url';
+import type { Advert } from './Advert';
 import { getAdvertById } from './get-advert-by-id';
 import { enableLazyLoad } from './lazy-load';
+import { memoizedFetchNonRefreshableLineItemIds } from './non-refreshable-line-items';
 
-const setSlotAdRefresh = (event: googletag.events.Event): void => {
+const outstreamSizes = [
+	adSizes.outstreamDesktop.toString(),
+	adSizes.outstreamMobile.toString(),
+	adSizes.outstreamGoogleDesktop.toString(),
+];
+
+/**
+ * Determine whether an advert should refresh, taking into account
+ * its size, whether there's a pageskin or whether the advert's
+ * line item is marked as non-refreshable
+ *
+ *  - Fluid ads should not refresh
+ *  - Outstream ads should not refresh
+ *  - Pageskins should not refresh
+ *  - Ads that have line items marked as non-refreshable should not be
+ * 	  refreshed. This information is retrieved via the DFP non refreshable
+ * 	  line item API endpoint
+ *
+ * @param advert The candidate advert to check
+ * @param nonRefreshableLineItemIds The array of line item ids for which
+ * adverts should not refresh
+ */
+const shouldRefresh = (
+	advert: Advert,
+	nonRefreshableLineItemIds: number[] = [],
+): boolean => {
+	const sizeString = advert.size?.toString();
+	const isNotFluid = sizeString !== '0,0';
+	const isOutstream =
+		sizeString && outstreamSizes.includes(sizeString as AdSizeString);
+	const isNonRefreshableLineItem =
+		advert.lineItemId &&
+		nonRefreshableLineItemIds.includes(advert.lineItemId);
+
+	return (
+		isNotFluid &&
+		!isOutstream &&
+		!config.get('page.hasPageSkin') &&
+		!isNonRefreshableLineItem
+	);
+};
+
+const setSlotAdRefresh = (
+	event: googletag.events.ImpressionViewableEvent,
+): void => {
 	const advert = getAdvertById(event.slot.getSlotElementId());
 	const viewabilityThresholdMs = 30000; // 30 seconds refresh
 
-	if (advert?.shouldRefresh) {
-		const onDocumentVisible = () => {
-			if (!document.hidden) {
-				document.removeEventListener(
-					'visibilitychange',
-					onDocumentVisible,
-				);
-				enableLazyLoad(advert);
-			}
-		};
-
-		setTimeout(() => {
-			// During the elapsed time, a 'disable-refresh' message may have been posted.
-			// Check the flag again.
-			if (!advert.shouldRefresh) {
-				return;
-			}
-			if (document.hidden) {
-				document.addEventListener(
-					'visibilitychange',
-					onDocumentVisible,
+	if (!advert) {
+		return;
+	}
+
+	// Asynchronously retrieve the non-refreshable line item ids
+	// Only do this if they haven't been attached to the page config
+	if (!window.guardian.config.page.dfpNonRefreshableLineItemIds) {
```

Comment:
What happens if the ids are already attached to the page config?

Reply:
If the ids have been attached to the page config then the user isn't in the variant of the experiment. Therefore the line items will be available at the `on-slot-render` point, and refreshing is checked in the usual way there.

This check is kind of a proxy for being in the variant of the experiment, I'd be interested in whether you think this is confusing and an additional check for being in the variant of the test would be helpful (even though both conditions would always simultaneously be true).

Reply:
I first got confused with this being a proxy, and I think this highlights a maintainability issue.

The main problem here is that once we’ve run our test and made a final decision on whether to use this approach, we won’t remember to remove this check, as there’s no direct link between the presence of `config.pagedfpNonRefreshableLineItemIds` and the value of `config.tests.fetchNonRefreshableLineItemsVariant`.

I’m all for exposing the server-side test client-side and making our decision based of that instead.

Reply:
> I first got confused with this being a proxy, and I think this highlights a maintainability issue.

Good point, if it's not clear then it can be improved. I'll have a look at exposing the check client-side and make the control flow explicitly based on test participation.